File: app/src/main/java/com/doapps/habits/fragments/ProgramFragment.java
package com.doapps.habits.fragments;

import android.annotation.SuppressLint;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.doapps.habits.BuildConfig;
import com.doapps.habits.R;
import com.doapps.habits.converters.IntegerArrayConverter;
import com.doapps.habits.dao.HabitDao;
import com.doapps.habits.database.ProgramsDatabase;
import com.doapps.habits.helper.HabitListManager;
import com.doapps.habits.models.Achievement;
import com.doapps.habits.models.Habit;
import com.doapps.habits.models.Program;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.List;

public class ProgramFragment extends Fragment {

  public static ProgramsDatabase database;
  public static HabitDao habitDatabase;
  private DatabaseReference programRef;

  private static List<Achievement> createAchievementList(DataSnapshot dataSnapshot) {
    List<Achievement> achievements = new ArrayList<>((int) dataSnapshot.getChildrenCount());
    for (DataSnapshot achievementSnapshot : dataSnapshot.getChildren()) {
      List<String> templates =
          new ArrayList<>((int) achievementSnapshot.getChildrenCount());
      for (DataSnapshot templatesSnapshot : achievementSnapshot.child("templates").getChildren()) {
        templates.add(templatesSnapshot.child("name").getValue(String.class));
      }
      Achievement achievement = new Achievement(
          achievementSnapshot.child("rating").getValue(Integer.class),
          templates);
      achievements.add(achievement);
    }
    return achievements;
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    View result = inflater.inflate(R.layout.fragment_program, container, false);

    FloatingActionButton fab = result.findViewById(R.id.fab);
    TextView description = result.findViewById(R.id.description);
    database = HabitListManager.getInstance(getContext()).getProgramDatabase();
    habitDatabase = HabitListManager.getInstance(getContext()).getDatabase().habitDao();

    int position = getArguments().getInt("pos", 0);
    programRef = FirebaseDatabase.getInstance().getReference().child("programs")
        .child(Integer.toString(position));
    programRef.addListenerForSingleValueEvent(new ValueEventListener() {
      @Override
      public void onDataChange(DataSnapshot dataSnapshot) {
        description
            .setText(dataSnapshot.child("habit").child("description").getValue(String.class));

        if (BuildConfig.DEBUG) {
          Log.i("Program fragment", dataSnapshot.toString());
        }

        if (fab != null) {
          fab.setOnClickListener(view -> new InsertIfNotExists(dataSnapshot).execute());
        }

      }

      @Override
      public void onCancelled(DatabaseError databaseError) {
        Log.e("onCancelled", databaseError.getMessage());
        getActivity().getSupportFragmentManager().beginTransaction().remove(ProgramFragment.this)
            .commit();
      }
    });
    return result;
  }

  private static class InsertTask extends AsyncTask<Program, Void, Void> {

    @Override
    protected Void doInBackground(Program... programs) {
      Log.i("DatabaseContains", "New program added");
      database.programDao().insertAll(programs);
      new DatabasePrintTask().execute();
      return null;
    }
  }

  private static class InsertIfNotExists extends AsyncTask<Void, Void, Void> {

    private final DataSnapshot mSnapshot;

    InsertIfNotExists(DataSnapshot snapshot) {
      mSnapshot = snapshot;
    }

    @Override
    protected Void doInBackground(Void... voids) {
      if (database.programDao().idExists(Integer.parseInt(mSnapshot.getKey())) == 0) {
        new InsertHabitTask(mSnapshot).execute();
      } else {
        Log.i("DatabaseContains", "Program already added");
      }
      return null;
    }
  }

  private static class DatabasePrintTask extends AsyncTask<Void, Void, Void> {

    @Override
    protected Void doInBackground(Void... voids) {
      Log.i("New Database: ", Arrays.toString(database.programDao().getAll().toArray()));
      return null;
    }
  }

  @SuppressLint("StaticFieldLeak")
  private static class InsertHabitTask extends AsyncTask<Void, Void, Long> {

    private final DataSnapshot dataSnapshot;

    InsertHabitTask(DataSnapshot dataSnapshot) {
      this.dataSnapshot = dataSnapshot;
    }

    @Override
    protected Long doInBackground(Void... voids) {
      Log.i("DatabaseContains", "New program added");
      Calendar upd = Calendar.getInstance();
      Habit habit = new Habit(
          dataSnapshot.child("habit").child("title").getValue(String.class),
          dataSnapshot.child("habit").child("question").getValue(String.class),
          false,
          upd.get(Calendar.DATE),
          upd.get(Calendar.MONTH),
          upd.get(Calendar.YEAR),
          dataSnapshot.child("habit").child("time").getValue(Integer.class),
          System.currentTimeMillis(),
          dataSnapshot.child("habit").child("cost").getValue(Integer.class),
          IntegerArrayConverter.fromArray(
              dataSnapshot.child("habit").child("frequency").getValue(String.class)));
      return habitDatabase.insert(habit);
    }

    @Override
    protected void onPostExecute(Long id) {
      super.onPostExecute(id);
      List<Achievement> achievements = createAchievementList(dataSnapshot.child("achievements"));

      Program program = new Program(
          Integer.valueOf(dataSnapshot.getKey()),
          dataSnapshot.child("name").getValue(String.class),
          String.format("%s SUCCESS", dataSnapshot.child("success").getValue(String.class)),
          id,
          achievements
      );
      new InsertTask().execute(program);
    }
  }
}

File: app/src/main/java/com/doapps/habits/fragments/ProgramFragment.kt
package com.doapps.habits.fragments

import android.annotation.SuppressLint
import android.content.Context
import android.os.AsyncTask
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import com.doapps.habits.BuildConfig
import com.doapps.habits.R
import com.doapps.habits.converters.IntegerArrayConverter
import com.doapps.habits.database.HabitsDatabase
import com.doapps.habits.database.ProgramsDatabase
import com.doapps.habits.helper.HabitListManager
import com.doapps.habits.models.Achievement
import com.doapps.habits.models.Habit
import com.doapps.habits.models.Program
import com.doapps.habits.receivers.NotificationReceiver
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import java.util.*
import java.util.concurrent.ExecutionException

class ProgramFragment : Fragment() {

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                            savedInstanceState: Bundle?): View? {
    // Inflate the layout for this fragment
    val result = inflater.inflate(R.layout.fragment_program, container, false)

    val fab = result.findViewById<FloatingActionButton>(R.id.fab)
    val description = result.findViewById<TextView>(R.id.description)
    val habitListManager = HabitListManager.getInstance(context)
    programDatabase = habitListManager.programDatabase
    habitDatabase = habitListManager.habitsDatabase

    val position = arguments!!.getInt("pos", 0)
    val programRef = FirebaseDatabase.getInstance().reference.child("programs")
        .child(Integer.toString(position))
    programRef.child("habit").child("description").addValueEventListener(object : ValueEventListener {
      override fun onDataChange(dataSnapshot: DataSnapshot) {
        description.text = dataSnapshot.getValue(String::class.java)
      }

      override fun onCancelled(databaseError: DatabaseError) {
        Log.e("onCancelled", databaseError.message)
      }
    })
    programRef.addValueEventListener(object : ValueEventListener {
      override fun onDataChange(dataSnapshot: DataSnapshot) {
        if (BuildConfig.DEBUG) {
          Log.i("Program fragment", dataSnapshot.toString())
        }

        fab.setOnClickListener {
          try {
            val isInserted = InsertIfNotExists(dataSnapshot, context!!).execute().get()
            if (!isInserted) {
              Toast.makeText(context!!.applicationContext, "Already added",
                  Toast.LENGTH_SHORT).show()
            }
          } catch (e: InterruptedException) {
            e.printStackTrace()
          } catch (e: ExecutionException) {
            e.printStackTrace()
          }
        }
      }

      override fun onCancelled(databaseError: DatabaseError) {
        Log.e("onCancelled", databaseError.message)
        activity!!.supportFragmentManager.beginTransaction()
            .remove(this@ProgramFragment)
            .commit()
      }
    })
    return result
  }

  private class InsertTask : AsyncTask<Program, Unit, Unit>() {

    override fun doInBackground(vararg programs: Program) {
      Log.i("DatabaseContains", "New program added")
      programDatabase.programDao().insertAll(*programs)
      DatabasePrintTask().execute()
    }
  }

  @SuppressLint("StaticFieldLeak")
  private class InsertIfNotExists internal constructor(private val mSnapshot: DataSnapshot, private val mContext: Context) : AsyncTask<Unit, Unit, Boolean>() {

    override fun doInBackground(vararg voids: Unit): Boolean? {
      val exists = programDatabase.programDao().idExists(Integer.parseInt(mSnapshot.key)) == 0
      if (exists) {
        InsertHabitTask(mSnapshot, mContext).execute()
      }
      return exists
    }
  }

  private class DatabasePrintTask : AsyncTask<Unit, Unit, Unit>() {

    override fun doInBackground(vararg voids: Unit) {
      Log.i("New Database", Arrays.toString(programDatabase.programDao().all.toTypedArray()))
    }
  }

  @SuppressLint("StaticFieldLeak")
  private class InsertHabitTask internal constructor(private val dataSnapshot: DataSnapshot, private val context: Context) : AsyncTask<Unit, Unit, Long>() {

    override fun doInBackground(vararg voids: Unit): Long? {
      Log.i("DatabaseContains", "New program added")
      val upd = Calendar.getInstance()
      val habit = Habit(
          dataSnapshot.child("habit").child("title").getValue(String::class.java),
          dataSnapshot.child("habit").child("question").getValue(String::class.java),
          false,
          upd.get(Calendar.DATE),
          upd.get(Calendar.MONTH),
          upd.get(Calendar.YEAR),
          dataSnapshot.child("habit").child("time").getValue(Int::class.java)!!,
          System.currentTimeMillis(),
          dataSnapshot.child("habit").child("cost").getValue(Int::class.java)!!,
          *IntegerArrayConverter
              .toArray(dataSnapshot.child("habit").child("frequency").getValue(String::class.java)))
      return habitDatabase.habitDao().insert(habit)
    }

    override fun onPostExecute(id: Long?) {
      super.onPostExecute(id)
      val notificationService = NotificationReceiver()
      notificationService.setAlarm(context, id!!,
          dataSnapshot.child("habit").child("question").getValue(String::class.java))

      val achievements = createAchievementList(dataSnapshot.child("achievements"))

      Log.i("ID", dataSnapshot.key)
      val program = Program(
          Integer.valueOf(dataSnapshot.key)!!,
          dataSnapshot.child("name").getValue(String::class.java),
          dataSnapshot.child("success").getValue(String::class.java),
          id,
          achievements
      )
      InsertTask().execute(program)
    }
  }

  companion object {

    private lateinit var programDatabase: ProgramsDatabase
    private lateinit var habitDatabase: HabitsDatabase


    private fun createAchievementList(dataSnapshot: DataSnapshot): List<Achievement> {
      val achievements = ArrayList<Achievement>(dataSnapshot.childrenCount.toInt())
      for (achievementSnapshot in dataSnapshot.children) {
        val templates = Array(achievementSnapshot.childrenCount.toInt(), { "" })
        templates.map {
          achievementSnapshot.child("templates").children.forEach {
            it.child("name").getValue(String::class.java)
          }
        }
        val achievement = Achievement(
            achievementSnapshot.child("rating").getValue(Int::class.java),
            templates)
        achievements.add(achievement)
      }
      return achievements
    }
  }
}
